Microseconds support (:type/UNIXTimestampMicroseconds) (#1889)

dpsutton opened the pull request:
fixes #1889 

Simple addition to support microseconds

Databases that have implementations for milliseconds also got an implementation of microseconds, presumably as there is a native way that is faster than division and converting seconds.

- oracle (seemingly just needs the literal 100000 marked as `hsql/raw`
- snowflake (uses a `to_timestamp expr 6` function)
- h2 natively understands the unit "microsecond" in its conversion function
- bigquery has a native `timestamp_micros` function

dpsutton committed:
Microseconds support (:type/UNIXTimestampMicroseconds) (#1889)

dpsutton committed:
Add mongo support and sqlite's different date formats in test

dpsutton committed:
Don't duplicate milliseconds defmethod for oracle

dpsutton committed:
Check for number not integer in unixtimestamp dispatch

oracle derives ultimately from :type/Decimal in this case which is not
a :type/Integer

sbelak commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -267,13 +271,24 @@
   (->honeysql driver expression-definition))
 
+(defn unix-timestamp->honeysql*
+  "Function in front of `unix-timestamp->honeysql` to translate the special type to the appropriate magnitude (seconds,
```
Minor nitpick: in the rest of the code we tend to use `unit` instead of `magnitude`

sbelak commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -267,13 +271,24 @@
   [driver [_ _ expression-definition]]
   (->honeysql driver expression-definition))
 
+(defn unix-timestamp->honeysql*
```
Star usually denotes an implementation function, while here it's a wrapper. Perhaps it would be cleaner to just package unit inference into own function and pass that to the OG `unix-timestamp->honeysql`. eg `special-type->unix-timestamp-unit`.

sbelak reviewed (approved):
Looks good. Minor naming nitpicks

dpsutton committed:
Translate special-type to unit in separate function

the one i had used was not well named (unix-timestamp->honeysql*) as
it implied an implementation function rather than a wrapper. But the
wrapper wasn't needed anyways

dpsutton committed:
Rename and docstring for `special-type->unix-timestamp-unit`

dpsutton commented:
Thanks @sbelak ! Updated with your thoughtful additions.
